### DISCUSSION TITLE
Cannot configuring DVC cache file link type as hardlink

### DISCUSSION THREAD

**Discussion by tshu-w** [treeverse/dvc#7509] [Help]
### Steps to reproduce
```console
mkdir dvc && cd dvc
git init && dvc init
dvc config cache.type hardlink
touch test.xml && dvc add test.xml
ls -l
```
got:
```
-rw-rw-r-- 1 wangtianshu wangtianshu  0 Mar 28 10:38 test.xml
-rw-rw-r-- 1 wangtianshu wangtianshu 73 Mar 28 10:38 test.xml.dvc
```
`test.xml` is copied to the cache instead of being hard-linked, even if the system supports it (verified by `ln` without `-s`). If I config cache type as soft link, it works fine. 

```console
dvc doctor
DVC version: 2.9.5 (pip)
---------------------------------
Platform: Python 3.10.2 on Linux-4.4.0-189-generic-x86_64-with-glibc2.31
Supports:
	webhdfs (fsspec = 2022.2.0),
	http (aiohttp = 3.8.1, aiohttp-retry = 2.4.6),
	https (aiohttp = 3.8.1, aiohttp-retry = 2.4.6),
	ssh (sshfs = 2021.11.2)
Cache types: hardlink, symlink
Cache directory: ext4 on /dev/sda1
Caches: local
Remotes: None
Workspace directory: ext4 on /dev/sda1
Repo: dvc, git
```

I can make hard link through `ln`:
```console
ln test.xml test2.xml && ls -l
```
got:
```
total 4
-rw-rw-r-- 2 wangtianshu wangtianshu  0 Mar 28 13:55 test2.xml
-rw-rw-r-- 2 wangtianshu wangtianshu  0 Mar 28 13:55 test.xml
-rw-rw-r-- 1 wangtianshu wangtianshu 73 Mar 28 13:54 test.xml.dvc
```

**Comment by pmrowla** [comment]
In this example case it is because you ran `dvc add` before changing the cache link type to hardlink (so the file was copied on `dvc add`). DVC will not re-link any existing files automatically when you set `cache.type.`

If you change the cache link type configuration, you need to run `dvc checkout --relink` if you want to force relinking. https://dvc.org/doc/command-reference/checkout#options

**Reply by tshu-w** [reply]
@pmrowla Sorry, I fix my commands order and make sure `dvc config` before `dvc add` and the issue is still there. I also run `dvc checkout --relink` but the text.xml is still not linked as hardlink.

**Reply by skshetry** [reply]
@tshu-w, can you please share the output of `dvc doctor`?

**Reply by tshu-w** [reply]
> @tshu-w, can you please share the output of `dvc doctor`?

```console
DVC version: 2.9.5 (pip)
---------------------------------
Platform: Python 3.10.2 on Linux-4.4.0-189-generic-x86_64-with-glibc2.31
Supports:
	webhdfs (fsspec = 2022.2.0),
	http (aiohttp = 3.8.1, aiohttp-retry = 2.4.6),
	https (aiohttp = 3.8.1, aiohttp-retry = 2.4.6),
	ssh (sshfs = 2021.11.2)
Cache types: hardlink, symlink
Cache directory: ext4 on /dev/sda1
Caches: local
Remotes: None
Workspace directory: ext4 on /dev/sda1
Repo: dvc, git
```

**Reply by skshetry** [reply]
> > @tshu-w, can you please share the output of `dvc doctor`?
> 
> ```
> DVC version: 2.9.5 (pip)
> ---------------------------------
> Platform: Python 3.10.2 on Linux-4.4.0-189-generic-x86_64-with-glibc2.31
> Supports:
> 	webhdfs (fsspec = 2022.2.0),
> 	http (aiohttp = 3.8.1, aiohttp-retry = 2.4.6),
> 	https (aiohttp = 3.8.1, aiohttp-retry = 2.4.6),
> 	ssh (sshfs = 2021.11.2)
> ``` 

This does not have repo information. Do you mind `cd`-ing into the Repo and then invoking `dvc doctor`? It has more detailed information. 🙂

**Reply by tshu-w** [reply]
Sorry, update it.

**Reply by dberenbaum** [reply]
@pmrowla @skshetry Any thoughts on this one?
